out_stackdriver: add 'k8s_container' , 'k8s_node' and 'k8s_pod' to supported resource types

JeffLuoo opened the pull request:
<!-- Provide summary of changes -->
For now stackdriver output plugin only supports 'global' and 'gce_instance' resource types. More resource types need to be supported like 'k8s_container", "k8s_node" and "k8s_pod"

This patch will enable fluent-bit stackdriver to support `k8s_container`, `k8s_node` and `k8s_pod` resource types.
<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, #77" -->
Issue number #1003 
----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change
<!-- Invoke Fluent Bit and Valgrind as: $ valgrind ./bin/fluent-bit <args> -->
- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [x] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

JeffLuoo commented:
## Example configuration file for the change
### k8s_container
fluent-bit.conf
```yaml
[SERVICE]
    Flush        1
    Daemon       Off
    Log_Level    debug
    Parsers_File myparser.conf

[INPUT]
    Name tail
    Path /tmp/testLog.json
    Parser JSON
    Tag k8s_container.namespace_name_test.pod_name_test.test_container_name

[OUTPUT]
    Name stackdriver
    Match *
    resource k8s_container
    k8s_cluster_name k8s_cluster_name_test
    k8s_cluster_location k8s_cluster_location_test

[FILTER]
    Name record_modifier
    Match k8s_container.*
    Record logging.googleapis.com/local_resource_id k8s_container.namespace_name_test.podt_name_test.container_name_test
```
myparser.conf
```yaml
[PARSER]
   Name  JSON
   Format  json
```
/tmp/testLog.json
```json
{"stream":"stderr","time":"2019-03-09T13:52:36.806411254Z","log":"test Log Message","kubernetes":{"labels":{"controller-revision-hash":"foorbarr","pod-template-generation":"1","app":"logging-availability-test"},"namespace_name":"test-components","docker_id":"testDockerId","container_name":"logging-test","pod_name":"logging-availability-test-v8gwl","annotations":{"prometheus.io\/path":"\/","prometheus.io\/scrape":"true"},"host":"gke-test-host","pod_id":"test-pod-id"}}
```

### k8s_node
fluent-bit.conf
```yaml
[SERVICE]
    Flush        1
    Daemon       Off
    Log_Level    debug
    Parsers_File myparser.conf

[INPUT]
    Name            systemd
    Tag             k8s_node.test
    Systemd_Filter  _SYSTEMD_UNIT=docker.service
    Parser          JSON
    Path            /run/log/journal/

[OUTPUT]
    Name stackdriver
    Match k8s_node.*
    Resource k8s_node
    k8s_cluster_name test_cluster_name
    k8s_cluster_location  test_cluster_lcation

[FILTER]
    Name record_modifier
    Match k8s_node.*
    Record logging.googleapis.com/local_resource_id k8s_node.testNodeName
```
myparser.conf
```yaml
[PARSER]
   Name  JSON
   Format  json
```
### k8s_pod
testPod.conf
```yaml
[SERVICE]
    Flush        1
    Daemon       Off
    Log_Level    debug
    Parsers_File myparser.conf

[INPUT]
    Name            tail
    Tag             k8s_pod.test
    Parser          JSON
    Path            /home/jeffluoo/fluentbit/testPod.json

[OUTPUT]
    Name stackdriver
    Match k8s_pod.*
    Resource k8s_pod
    k8s_cluster_name test_cluster_name
    k8s_cluster_location  test_cluster_lcation
    service_account_email jeffluoo@google.com
    google_service_credentials /home/jeffluoo/fluentbit/gke-config.json

[FILTER]
    Name record_modifier
    Match k8s_pod.*
    Record logging.googleapis.com/local_resource_id k8s_pod.test1.test2
```
myparser.conf
```yaml
[PARSER]
   Name  JSON
   Format  json
```
testPod.json
```json
{"log":"time=\"2020-06-10T01:19:04Z\" level=info msg=\"Listening on 127.0.0.1:9090\" source=\"node_exporter.go:170\"\n","stream":"stderr","time":"2020-06-10T01:19:04.939109232Z"}
```

JeffLuoo commented:
##  Debug log output from testing the change
### k8s_container
```
fluent-bit/build$ ./bin/fluent-bit -c ../conf/fluent-bit.conf 
Fluent Bit v1.5.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/06/09 10:21:51] [ info] Configuration:
[2020/06/09 10:21:51] [ info]  flush time     | 1.000000 seconds
[2020/06/09 10:21:51] [ info]  grace          | 5 seconds
[2020/06/09 10:21:51] [ info]  daemon         | 0
[2020/06/09 10:21:51] [ info] ___________
[2020/06/09 10:21:51] [ info]  inputs:
[2020/06/09 10:21:51] [ info]      tail
[2020/06/09 10:21:51] [ info] ___________
[2020/06/09 10:21:51] [ info]  filters:
[2020/06/09 10:21:51] [ info]      record_modifier.0
[2020/06/09 10:21:51] [ info] ___________
[2020/06/09 10:21:51] [ info]  outputs:
[2020/06/09 10:21:51] [ info]      stackdriver.0
[2020/06/09 10:21:51] [ info] ___________
[2020/06/09 10:21:51] [ info]  collectors:
[2020/06/09 10:21:51] [debug] [storage] [cio stream] new stream registered: tail.0
[2020/06/09 10:21:51] [ info] [storage] version=1.0.4, initializing...
[2020/06/09 10:21:51] [ info] [storage] in-memory
[2020/06/09 10:21:51] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/06/09 10:21:51] [ info] [engine] started (pid=3033)
[2020/06/09 10:21:51] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/06/09 10:21:51] [debug] [input:tail:tail.0] inotify watch fd=19
[2020/06/09 10:21:51] [debug] [input:tail:tail.0] scanning path /tmp/testLog.json
[2020/06/09 10:21:51] [debug] [input:tail:tail.0] inode=829615 appended as /tmp/testLog.json
[2020/06/09 10:21:51] [debug] [input:tail:tail.0] scan_glob add(): /tmp/testLog.json, inode 829615
[2020/06/09 10:21:52] [ info] [oauth2] HTTP Status=200
[2020/06/09 10:21:52] [ info] [oauth2] access token from 'www.googleapis.com:443' retrieved
[2020/06/09 10:21:52] [debug] [router] match rule tail.0:stackdriver.0
[2020/06/09 10:21:52] [ info] [sp] stream processor started
[2020/06/09 10:21:52] [debug] [input:tail:tail.0] inode=829615 file=/tmp/testLog.json promote to TAIL_EVENT
[2020/06/09 10:21:52] [ info] inotify_fs_add(): inode=829615 watch_fd=1 name=/tmp/testLog.json
[2020/06/09 10:21:52] [debug] [task] created task=0x589a763b12d0 id=0 OK
[2020/06/09 10:21:52] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/06/09 10:21:52] [debug] [upstream] KA connection #30 to logging.googleapis.com:443 is now available
[2020/06/09 10:21:52] [debug] [task] destroy task=0x589a763b12d0 (task_id=0)
[2020/06/09 10:22:22] [debug] [upstream] drop keepalive connection to logging.googleapis.com:443 (keepalive idle timeout)
[2020/06/09 10:22:22] [debug] [upstream] KA connection #30 to logging.googleapis.com:443 has been disconnected by the remote service
^C[engine] caught signal (SIGINT)
[2020/06/09 10:22:33] [ info] [input] pausing tail.0
[2020/06/09 10:22:33] [debug] [input:tail:tail.0] inode=829615 removing file name /tmp/testLog.json
[2020/06/09 10:22:33] [ info] inotify_fs_remove(): inode=829615 watch_fd=1
```
### k8s_node
```
../build/bin/fluent-bit -c fluent-bit.conf 
Fluent Bit v1.5.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/06/10 09:54:15] [ info] Configuration:
[2020/06/10 09:54:15] [ info]  flush time     | 1.000000 seconds
[2020/06/10 09:54:15] [ info]  grace          | 5 seconds
[2020/06/10 09:54:15] [ info]  daemon         | 0
[2020/06/10 09:54:15] [ info] ___________
[2020/06/10 09:54:15] [ info]  inputs:
[2020/06/10 09:54:15] [ info]      systemd
[2020/06/10 09:54:15] [ info] ___________
[2020/06/10 09:54:15] [ info]  filters:
[2020/06/10 09:54:15] [ info]      record_modifier.0
[2020/06/10 09:54:15] [ info] ___________
[2020/06/10 09:54:15] [ info]  outputs:
[2020/06/10 09:54:15] [ info]      stackdriver.0
[2020/06/10 09:54:15] [ info] ___________
[2020/06/10 09:54:15] [ info]  collectors:
[2020/06/10 09:54:15] [debug] [storage] [cio stream] new stream registered: systemd.0
[2020/06/10 09:54:15] [ info] [storage] version=1.0.4, initializing...
[2020/06/10 09:54:15] [ info] [storage] in-memory
[2020/06/10 09:54:15] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/06/10 09:54:15] [ info] [engine] started (pid=22967)
[2020/06/10 09:54:15] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/06/10 09:54:15] [debug] [input:systemd:systemd.0] add filter: _SYSTEMD_UNIT=docker.service (or)
[2020/06/10 09:54:15] [debug] [input:systemd:systemd.0] sd_journal library may truncate values to sd_journal_get_data_threshold() bytes: 65536
[2020/06/10 09:54:15] [ info] [oauth2] HTTP Status=200
[2020/06/10 09:54:15] [ info] [oauth2] access token from 'www.googleapis.com:443' retrieved
[2020/06/10 09:54:15] [debug] [router] match rule systemd.0:stackdriver.0
[2020/06/10 09:54:15] [ info] [sp] stream processor started
[2020/06/10 09:54:15] [debug] [task] created task=0x58948bd59c30 id=0 OK
[2020/06/10 09:54:16] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/06/10 09:54:16] [debug] [upstream] KA connection #28 to logging.googleapis.com:443 is now available
[2020/06/10 09:54:16] [debug] [task] destroy task=0x58948bd59c30 (task_id=0)
^C[engine] caught signal (SIGINT)
[2020/06/10 09:54:44] [ info] [input] pausing systemd.0
```
### k8s_pod
```
./bin/fluent-bit -c ../conf/testPod.conf 
Fluent Bit v1.5.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/06/12 08:21:16] [ info] Configuration:
[2020/06/12 08:21:16] [ info]  flush time     | 1.000000 seconds
[2020/06/12 08:21:16] [ info]  grace          | 5 seconds
[2020/06/12 08:21:16] [ info]  daemon         | 0
[2020/06/12 08:21:16] [ info] ___________
[2020/06/12 08:21:16] [ info]  inputs:
[2020/06/12 08:21:16] [ info]      tail
[2020/06/12 08:21:16] [ info] ___________
[2020/06/12 08:21:16] [ info]  filters:
[2020/06/12 08:21:16] [ info]      record_modifier.0
[2020/06/12 08:21:16] [ info] ___________
[2020/06/12 08:21:16] [ info]  outputs:
[2020/06/12 08:21:16] [ info]      stackdriver.0
[2020/06/12 08:21:16] [ info] ___________
[2020/06/12 08:21:16] [ info]  collectors:
[2020/06/12 08:21:16] [debug] [storage] [cio stream] new stream registered: tail.0
[2020/06/12 08:21:16] [ info] [storage] version=1.0.4, initializing...
[2020/06/12 08:21:16] [ info] [storage] in-memory
[2020/06/12 08:21:16] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/06/12 08:21:16] [ info] [engine] started (pid=27495)
[2020/06/12 08:21:16] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/06/12 08:21:16] [debug] [input:tail:tail.0] inotify watch fd=19
[2020/06/12 08:21:16] [debug] [input:tail:tail.0] scanning path /home/jeffluoo/fluentbit/testPod.json
[2020/06/12 08:21:16] [debug] [input:tail:tail.0] inode=965796 appended as /home/jeffluoo/fluentbit/testPod.json
[2020/06/12 08:21:16] [debug] [input:tail:tail.0] scan_glob add(): /home/jeffluoo/fluentbit/testPod.json, inode 965796
[2020/06/12 08:21:16] [ info] [oauth2] HTTP Status=200
[2020/06/12 08:21:16] [ info] [oauth2] access token from 'www.googleapis.com:443' retrieved
[2020/06/12 08:21:16] [debug] [router] match rule tail.0:stackdriver.0
[2020/06/12 08:21:16] [ info] [sp] stream processor started
[2020/06/12 08:21:16] [debug] [input:tail:tail.0] inode=965796 file=/home/jeffluoo/fluentbit/testPod.json promote to TAIL_EVENT
[2020/06/12 08:21:16] [ info] inotify_fs_add(): inode=965796 watch_fd=1 name=/home/jeffluoo/fluentbit/testPod.json
[2020/06/12 08:21:17] [debug] [task] created task=0x55bf1b6af3e0 id=0 OK
[2020/06/12 08:21:17] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/06/12 08:21:17] [debug] [upstream] KA connection #30 to logging.googleapis.com:443 is now available
[2020/06/12 08:21:17] [debug] [task] destroy task=0x55bf1b6af3e0 (task_id=0)
^C[engine] caught signal (SIGINT)
[2020/06/12 08:21:24] [ info] [input] pausing tail.0
[2020/06/12 08:21:24] [debug] [input:tail:tail.0] inode=965796 removing file name /home/jeffluoo/fluentbit/testPod.json
[2020/06/12 08:21:24] [ info] inotify_fs_remove(): inode=965796 watch_fd=1
```

JeffLuoo commented:
##  Attached Valgrind output that shows no leaks or memory corruption was found
### k8s_container
```
$ valgrind ./bin/fluent-bit -c ../conf/fluent-bit.conf 
==3463== Memcheck, a memory error detector
==3463== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==3463== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==3463== Command: ./bin/fluent-bit -c ../conf/fluent-bit.conf
==3463== 
Fluent Bit v1.5.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/06/09 13:35:03] [ info] Configuration:
[2020/06/09 13:35:03] [ info]  flush time     | 1.000000 seconds
[2020/06/09 13:35:03] [ info]  grace          | 5 seconds
[2020/06/09 13:35:03] [ info]  daemon         | 0
[2020/06/09 13:35:03] [ info] ___________
[2020/06/09 13:35:03] [ info]  inputs:
[2020/06/09 13:35:03] [ info]      tail
[2020/06/09 13:35:03] [ info] ___________
[2020/06/09 13:35:03] [ info]  filters:
[2020/06/09 13:35:03] [ info]      record_modifier.0
[2020/06/09 13:35:03] [ info] ___________
[2020/06/09 13:35:03] [ info]  outputs:
[2020/06/09 13:35:03] [ info]      stackdriver.0
[2020/06/09 13:35:03] [ info] ___________
[2020/06/09 13:35:03] [ info]  collectors:
[2020/06/09 13:35:03] [debug] [storage] [cio stream] new stream registered: tail.0
[2020/06/09 13:35:03] [ info] [storage] version=1.0.4, initializing...
[2020/06/09 13:35:03] [ info] [storage] in-memory
[2020/06/09 13:35:03] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/06/09 13:35:03] [ info] [engine] started (pid=3463)
[2020/06/09 13:35:03] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/06/09 13:35:03] [debug] [input:tail:tail.0] inotify watch fd=19
[2020/06/09 13:35:03] [debug] [input:tail:tail.0] scanning path /tmp/testLog.json
[2020/06/09 13:35:03] [debug] [input:tail:tail.0] inode=829615 appended as /tmp/testLog.json
[2020/06/09 13:35:03] [debug] [input:tail:tail.0] scan_glob add(): /tmp/testLog.json, inode 829615
[2020/06/09 13:35:06] [ info] [oauth2] access token from 'www.googleapis.com:443' retrieved
[2020/06/09 13:35:06] [debug] [router] match rule tail.0:stackdriver.0
[2020/06/09 13:35:06] [ info] [sp] stream processor started
[2020/06/09 13:35:06] [debug] [input:tail:tail.0] inode=829615 file=/tmp/testLog.json promote to TAIL_EVENT
[2020/06/09 13:35:06] [ info] inotify_fs_add(): inode=829615 watch_fd=1 name=/tmp/testLog.json
==3463== Warning: client switching stacks?  SP change: 0x1ffefffcb8 --> 0x5f3fbf0
==3463==          to suppress, use: --max-stackframe=137322299592 or greater
==3463== Warning: client switching stacks?  SP change: 0x5f3fb68 --> 0x1ffefffcb8
==3463==          to suppress, use: --max-stackframe=137322299728 or greater
==3463== Warning: client switching stacks?  SP change: 0x1ffefffcb8 --> 0x5f3fb68
==3463==          to suppress, use: --max-stackframe=137322299728 or greater
==3463==          further instances of this message will not be shown.
[2020/06/09 13:35:07] [debug] [task] created task=0x5f39790 id=0 OK
[2020/06/09 13:35:08] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/06/09 13:35:08] [debug] [upstream] KA connection #30 to logging.googleapis.com:443 is now available
[2020/06/09 13:35:08] [debug] [task] destroy task=0x5f39790 (task_id=0)
^C[engine] caught signal (SIGINT)
[2020/06/09 13:35:15] [ info] [input] pausing tail.0
[2020/06/09 13:35:15] [debug] [input:tail:tail.0] inode=829615 removing file name /tmp/testLog.json
[2020/06/09 13:35:15] [ info] inotify_fs_remove(): inode=829615 watch_fd=1
==3463== 
==3463== HEAP SUMMARY:
==3463==     in use at exit: 0 bytes in 0 blocks
==3463==   total heap usage: 83,087 allocs, 83,087 frees, 10,339,378 bytes allocated
==3463== 
==3463== All heap blocks were freed -- no leaks are possible
==3463== 
==3463== For counts of detected and suppressed errors, rerun with: -v
==3463== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
### k8s_node
```
valgrind ../build/bin/fluent-bit -c fluent-bit.conf 
==22971== Memcheck, a memory error detector
==22971== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==22971== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==22971== Command: ../build/bin/fluent-bit -c fluent-bit.conf
==22971== 
Fluent Bit v1.5.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/06/10 09:59:34] [ info] Configuration:
[2020/06/10 09:59:34] [ info]  flush time     | 1.000000 seconds
[2020/06/10 09:59:34] [ info]  grace          | 5 seconds
[2020/06/10 09:59:34] [ info]  daemon         | 0
[2020/06/10 09:59:34] [ info] ___________
[2020/06/10 09:59:34] [ info]  inputs:
[2020/06/10 09:59:34] [ info]      systemd
[2020/06/10 09:59:34] [ info] ___________
[2020/06/10 09:59:34] [ info]  filters:
[2020/06/10 09:59:34] [ info]      record_modifier.0
[2020/06/10 09:59:34] [ info] ___________
[2020/06/10 09:59:34] [ info]  outputs:
[2020/06/10 09:59:34] [ info]      stackdriver.0
[2020/06/10 09:59:34] [ info] ___________
[2020/06/10 09:59:34] [ info]  collectors:
[2020/06/10 09:59:34] [debug] [storage] [cio stream] new stream registered: systemd.0
[2020/06/10 09:59:34] [ info] [storage] version=1.0.4, initializing...
[2020/06/10 09:59:34] [ info] [storage] in-memory
[2020/06/10 09:59:34] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/06/10 09:59:34] [ info] [engine] started (pid=22971)
[2020/06/10 09:59:34] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/06/10 09:59:34] [debug] [input:systemd:systemd.0] add filter: _SYSTEMD_UNIT=docker.service (or)
[2020/06/10 09:59:34] [debug] [input:systemd:systemd.0] sd_journal library may truncate values to sd_journal_get_data_threshold() bytes: 65536
[2020/06/10 09:59:36] [ info] [oauth2] HTTP Status=200
[2020/06/10 09:59:36] [ info] [oauth2] access token from 'www.googleapis.com:443' retrieved
[2020/06/10 09:59:36] [debug] [router] match rule systemd.0:stackdriver.0
[2020/06/10 09:59:36] [ info] [sp] stream processor started
==22971== Warning: client switching stacks?  SP change: 0x1fff0000e8 --> 0x81a72a0
==22971==          to suppress, use: --max-stackframe=137286225480 or greater
==22971== Warning: client switching stacks?  SP change: 0x81a7218 --> 0x1fff0000e8
==22971==          to suppress, use: --max-stackframe=137286225616 or greater
==22971== Warning: client switching stacks?  SP change: 0x1fff0000e8 --> 0x81a7218
==22971==          to suppress, use: --max-stackframe=137286225616 or greater
==22971==          further instances of this message will not be shown.
[2020/06/10 09:59:37] [debug] [task] created task=0x81a0e80 id=0 OK
[2020/06/10 09:59:37] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/06/10 09:59:37] [debug] [upstream] KA connection #28 to logging.googleapis.com:443 is now available
[2020/06/10 09:59:37] [debug] [task] destroy task=0x81a0e80 (task_id=0)
^C[engine] caught signal (SIGINT)
[2020/06/10 09:59:49] [ info] [input] pausing systemd.0
==22971== 
==22971== HEAP SUMMARY:
==22971==     in use at exit: 0 bytes in 0 blocks
==22971==   total heap usage: 83,113 allocs, 83,113 frees, 10,857,748 bytes allocated
==22971== 
==22971== All heap blocks were freed -- no leaks are possible
==22971== 
==22971== For counts of detected and suppressed errors, rerun with: -v
==22971== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
### k8s_pod
```
valgrind --leak-check=full ./bin/fluent-bit -c ../conf/testPod.conf
==27504== Memcheck, a memory error detector
==27504== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==27504== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==27504== Command: ./bin/fluent-bit -c ../conf/testPod.conf
==27504== 
Fluent Bit v1.5.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/06/12 08:23:10] [ info] Configuration:
[2020/06/12 08:23:10] [ info]  flush time     | 1.000000 seconds
[2020/06/12 08:23:10] [ info]  grace          | 5 seconds
[2020/06/12 08:23:10] [ info]  daemon         | 0
[2020/06/12 08:23:10] [ info] ___________
[2020/06/12 08:23:10] [ info]  inputs:
[2020/06/12 08:23:10] [ info]      tail
[2020/06/12 08:23:10] [ info] ___________
[2020/06/12 08:23:10] [ info]  filters:
[2020/06/12 08:23:10] [ info]      record_modifier.0
[2020/06/12 08:23:10] [ info] ___________
[2020/06/12 08:23:10] [ info]  outputs:
[2020/06/12 08:23:10] [ info]      stackdriver.0
[2020/06/12 08:23:10] [ info] ___________
[2020/06/12 08:23:10] [ info]  collectors:
[2020/06/12 08:23:10] [debug] [storage] [cio stream] new stream registered: tail.0
[2020/06/12 08:23:10] [ info] [storage] version=1.0.4, initializing...
[2020/06/12 08:23:10] [ info] [storage] in-memory
[2020/06/12 08:23:10] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/06/12 08:23:10] [ info] [engine] started (pid=27504)
[2020/06/12 08:23:10] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/06/12 08:23:10] [debug] [input:tail:tail.0] inotify watch fd=19
[2020/06/12 08:23:10] [debug] [input:tail:tail.0] scanning path /home/jeffluoo/fluentbit/testPod.json
[2020/06/12 08:23:10] [debug] [input:tail:tail.0] inode=965796 appended as /home/jeffluoo/fluentbit/testPod.json
[2020/06/12 08:23:10] [debug] [input:tail:tail.0] scan_glob add(): /home/jeffluoo/fluentbit/testPod.json, inode 965796
[2020/06/12 08:23:12] [ info] [oauth2] HTTP Status=200
[2020/06/12 08:23:12] [ info] [oauth2] access token from 'www.googleapis.com:443' retrieved
[2020/06/12 08:23:12] [debug] [router] match rule tail.0:stackdriver.0
[2020/06/12 08:23:12] [ info] [sp] stream processor started
[2020/06/12 08:23:12] [debug] [input:tail:tail.0] inode=965796 file=/home/jeffluoo/fluentbit/testPod.json promote to TAIL_EVENT
[2020/06/12 08:23:12] [ info] inotify_fs_add(): inode=965796 watch_fd=1 name=/home/jeffluoo/fluentbit/testPod.json
==27504== Warning: client switching stacks?  SP change: 0x1ffefffd38 --> 0x616ed10
==27504==          to suppress, use: --max-stackframe=137320009768 or greater
==27504== Warning: client switching stacks?  SP change: 0x616ec88 --> 0x1ffefffd38
==27504==          to suppress, use: --max-stackframe=137320009904 or greater
==27504== Warning: client switching stacks?  SP change: 0x1ffefffd38 --> 0x616ec88
==27504==          to suppress, use: --max-stackframe=137320009904 or greater
==27504==          further instances of this message will not be shown.
[2020/06/12 08:23:13] [debug] [task] created task=0x61688f0 id=0 OK
[2020/06/12 08:23:14] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/06/12 08:23:14] [debug] [upstream] KA connection #30 to logging.googleapis.com:443 is now available
[2020/06/12 08:23:14] [debug] [task] destroy task=0x61688f0 (task_id=0)
^C[engine] caught signal (SIGINT)
[2020/06/12 08:23:21] [ info] [input] pausing tail.0
[2020/06/12 08:23:21] [debug] [input:tail:tail.0] inode=965796 removing file name /home/jeffluoo/fluentbit/testPod.json
[2020/06/12 08:23:21] [ info] inotify_fs_remove(): inode=965796 watch_fd=1
==27504== 
==27504== HEAP SUMMARY:
==27504==     in use at exit: 0 bytes in 0 blocks
==27504==   total heap usage: 83,227 allocs, 83,227 frees, 10,342,569 bytes allocated
==27504== 
==27504== All heap blocks were freed -- no leaks are possible
==27504== 
==27504== For counts of detected and suppressed errors, rerun with: -v
==27504== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

igorpeshansky commented:
See also #2183. @heartrobotninja FYI.

JeffLuoo commented:
@igorpeshansky Thanks for the information.

StevenYCChou commented:
To be specific, I think adding incorporating tests are needed. See the comment: 
https://github.com/fluent/fluent-bit/pull/2183#issuecomment-645067551

This PR implemented the tests so it can be a reference too: https://github.com/fluent/fluent-bit/pull/2217

JeffLuoo commented:
@StevenYCChou Thanks for the information !! Tests are added in the new commit.

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+        if (k.via.str.size == key_size && 
+            strncmp(key, (char *) k.via.str.ptr, k.via.str.size) == 0) {
+            /* make sure to free it after use */ 
+            ptr =  flb_strndup(v.via.str.ptr, v.via.str.size);
```
use flb_sds_create_len() instead, mostly because in the caller you want to retrieve the string length, so having a flb_sds_t you can do something like flb_sds_len() which knows before hand the string length.

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+            return -1;
+        }
+        
+        ptr = strtok(local_resource_id, delim);
```
use of strtok(3) is discouraged since it might alter the original string buffer.

JeffLuoo replied:
Hi @edsiper , I am wondering that is it safe to use strtok(3) if the input local_resource_id is a copy of the original string? As flb_sds_create_len will call memcpy(). Or what you mean "it might alter the original string buffer" is that strtok will potentially messy up the area of memory that the input string is using?

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+        }
+        
+        ptr = strtok(local_resource_id, delim);
+        len_ptr = strlen(ptr);
```
avoid string byte counting, if you use a flb_sds_t you can know the string length before hand.

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+        if (strncmp(type, "k8s_container", strlen("k8s_container")) == 0) {
+            /* check the prefix */
+            if (strncasecmp(ptr, type, len_ptr) != 0) {
+                free(local_resource_id);
```
use flb_free() instead.

JeffLuoo replied:
switched to use flb_sds_t for local_resource_id so updated to use flb_sds_destroy

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+                return -1;
+            }
+
+            ptr = strtok(NULL, delim);
```
no strtok

JeffLuoo replied:
removed

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+            }
+
+            ptr = strtok(NULL, delim);
+            while (ptr != NULL)
```
align the brace on the right side:

https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+                    ctx->container_name = flb_sds_create(ptr);
+                }
+
+                ptr = strtok(NULL, delim);
```
no strtok

edsiper replied:
why a strtok(NULL, ..) ?

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+                return -1;
+            }
+        }
+        else if (strncmp(type, "k8s_node", strlen("k8s_node")) == 0) {
```
don't use strlen as a nested value. If the string you are comparing has a fixed value, there is no need to calculate it over and over.

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+                return -1;
+            }
+
+            ptr = strtok(NULL, delim);
```
no strtok

edsiper replied:
strtok(NULL..) ?

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+                return -1;
+            }
+        }
+        else if (strncmp(type, "k8s_pod", strlen("k8s_pod")) == 0) {
```
no strlen for fixed string

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+                return -1;
+            }
+
+            ptr = strtok(NULL, delim);
```
no strtok

JeffLuoo replied:
removed

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+            }
+
+            ptr = strtok(NULL, delim);
+            while (ptr != NULL)
```
fix brace

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+                    ctx->pod_name = flb_sds_create(ptr);
+                }
+
+                ptr = strtok(NULL, delim);
```
no strtok

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+        }
+
+        ret = 0;
+        free(local_resource_id);
```
use flb_free()

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -530,6 +699,149 @@ static int stackdriver_format(struct flb_config *config,
     }
+    else if (strcmp(ctx->resource, "k8s_container") == 0) {
+        /* k8s_container resource has fields project_id, location, cluster_name,
+                                             namespace_name, pod_name, container_name */
```
multiline comments: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#multiline-comment

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -530,6 +699,149 @@ static int stackdriver_format(struct flb_config *config,
+        /* k8s_container resource has fields project_id, location, cluster_name,
+                                             namespace_name, pod_name, container_name */
+
+        /* the local_resource_id for k8s_container is in format: 
```
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#multiline-comment

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -530,6 +699,149 @@ static int stackdriver_format(struct flb_config *config,
+        msgpack_pack_str_body(&mp_pck, "project_id", 10);
+        msgpack_pack_str(&mp_pck, flb_sds_len(ctx->project_id));
+        msgpack_pack_str_body(&mp_pck,
+                                ctx->project_id, flb_sds_len(ctx->project_id));
```
fix alignment

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -530,6 +699,149 @@ static int stackdriver_format(struct flb_config *config,
+        msgpack_pack_str_body(&mp_pck, "location", 8);
+        msgpack_pack_str(&mp_pck, flb_sds_len(ctx->cluster_location));
+        msgpack_pack_str_body(&mp_pck,
+                                ctx->cluster_location, flb_sds_len(ctx->cluster_location));
```
fix alignment

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -530,6 +699,149 @@ static int stackdriver_format(struct flb_config *config,
+        msgpack_pack_str_body(&mp_pck, "cluster_name", 12);
+        msgpack_pack_str(&mp_pck, flb_sds_len(ctx->cluster_name));
+        msgpack_pack_str_body(&mp_pck,
+                                ctx->cluster_name, flb_sds_len(ctx->cluster_name));
```
fix alignment

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -530,6 +699,149 @@ static int stackdriver_format(struct flb_config *config,
+        msgpack_pack_str_body(&mp_pck, "namespace_name", 14);
+        msgpack_pack_str(&mp_pck, flb_sds_len(ctx->namespace_name));
+        msgpack_pack_str_body(&mp_pck,
+                                ctx->namespace_name, flb_sds_len(ctx->namespace_name));
```
fix alignment

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -530,6 +699,149 @@ static int stackdriver_format(struct flb_config *config,
+        msgpack_pack_str_body(&mp_pck, "pod_name", 8);
+        msgpack_pack_str(&mp_pck, flb_sds_len(ctx->pod_name));
+        msgpack_pack_str_body(&mp_pck,
+                                ctx->pod_name, flb_sds_len(ctx->pod_name));
```
fix alignment

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -530,6 +699,149 @@ static int stackdriver_format(struct flb_config *config,
+        /* k8s_node resource has fields project_id, location, cluster_name, node_name */
+
+        /* the local_resource_id for k8s_node is in format: 
+            k8s_node.<node_name> */
```
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#multiline-comment

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -530,6 +699,149 @@ static int stackdriver_format(struct flb_config *config,
+        msgpack_pack_str_body(&mp_pck, "project_id", 10);
+        msgpack_pack_str(&mp_pck, flb_sds_len(ctx->project_id));
+        msgpack_pack_str_body(&mp_pck,
+                                ctx->project_id, flb_sds_len(ctx->project_id));
```
fix alignment

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -530,6 +699,149 @@ static int stackdriver_format(struct flb_config *config,
+        msgpack_pack_str_body(&mp_pck, "location", 8);
+        msgpack_pack_str(&mp_pck, flb_sds_len(ctx->cluster_location));
+        msgpack_pack_str_body(&mp_pck,
+                                ctx->cluster_location, flb_sds_len(ctx->cluster_location));
```
fix alignment

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -530,6 +699,149 @@ static int stackdriver_format(struct flb_config *config,
+        msgpack_pack_str_body(&mp_pck, "cluster_name", 12);
+        msgpack_pack_str(&mp_pck, flb_sds_len(ctx->cluster_name));
+        msgpack_pack_str_body(&mp_pck,
+                                ctx->cluster_name, flb_sds_len(ctx->cluster_name));
```
fix alignment

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -530,6 +699,149 @@ static int stackdriver_format(struct flb_config *config,
+        msgpack_pack_str_body(&mp_pck, "node_name", 9);
+        msgpack_pack_str(&mp_pck, flb_sds_len(ctx->node_name));
+        msgpack_pack_str_body(&mp_pck,
+                                ctx->node_name, flb_sds_len(ctx->node_name));
```
fix alignment

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -530,6 +699,149 @@ static int stackdriver_format(struct flb_config *config,
+        k8s_resource_type = 1;
+    }
+    else if (strcmp(ctx->resource, "k8s_pod") == 0) {
+        /* k8s_pod resource has fields project_id, location, cluster_name, 
```
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#multiline-comment

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -530,6 +699,149 @@ static int stackdriver_format(struct flb_config *config,
+        /* k8s_pod resource has fields project_id, location, cluster_name, 
+                                       namespace_name, pod_name */
+
+        /* the local_resource_id for k8s_pod is in format: 
```
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#multiline-comment

JeffLuoo replied:
updated

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -567,7 +880,35 @@ static int stackdriver_format(struct flb_config *config,
+
+            for (i = 0; i < map_size; ++i) {
+                kv = &obj->via.map.ptr[i];
+                if (strncmp(kv->key.via.str.ptr, LOCAL_RESOURCE_ID_KEY, len) == 0) {
```
to optimize you can check for the key length first and then do the strncmp()

JeffLuoo replied:
updated

edsiper commented:
@JeffLuoo thanks for your work on this, it looks good. I added some minor request changes

JeffLuoo commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+    msgpack_object v;
+    char *ptr = NULL;
+
+    key_size = strlen(key);
```
updated

JeffLuoo replied:
updated

JeffLuoo commented:
Hi @edsiper Thanks for taking time to review this PR. I have committed and pushed my most recent update to resolve most of comments. I replied to the comment for [strtok](https://github.com/fluent/fluent-bit/pull/2281/files#r445785625) as `memcpy` has been used.

Jeff

JeffLuoo commented:
Hi @edsiper Thanks for reminding me the function flb_slist_split_string() !! It is exactly what I am looking for. I have rebased and force pushed my update to resolve the strtok() comments and the merging conflicts on those file I modified. PTAL of my latest change. Thanks again for your time!

courageJ commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -278,6 +278,241 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+/* 
+*    process_local_resource_id():
+*  - extract the value from "logging.googleapis.com/local_resource_id" field
+*  - use extracted value to assign the label keys for differnet resource types
```
typo: differnet

JeffLuoo replied:
oops. updated

courageJ commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -278,6 +278,241 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+*    process_local_resource_id():
+*  - extract the value from "logging.googleapis.com/local_resource_id" field
+*  - use extracted value to assign the label keys for differnet resource types
+*    that specified in the configuration of stackdriver_out plugin
```
nit: that are specified?

JeffLuoo replied:
updated

courageJ commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -278,6 +278,241 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+                                                           LEN_LOCAL_RESOURCE_ID_KEY);
+        if (!local_resource_id) {
+            msgpack_unpacked_destroy(&result);
+            return -1;
```
do we need log the error/warning here?

JeffLuoo replied:
when process_local_resource_id returns, the caller function will catch the error and log the error/warning. WDYT?

edsiper commented on plugins/out_stackdriver/stackdriver.h:
```diff
@@ -47,6 +47,12 @@
+#define LOCAL_RESOURCE_ID_KEY "logging.googleapis.com/local_resource_id"
+#define LEN_LOCAL_RESOURCE_ID_KEY 40
+
+#define LEN_K8S_CONTAINER strlen("k8s_container")
```
the common way to do this is like:

```
#define K8S_CONTAINER    "k8s_container"
```

then in the places that you need to get the string length you just do:

```
len = sizeof(K8S_CONTAINER) - 1;
```

> the value of  ```sizeof()``` is set at compile time

JeffLuoo replied:
updated.

courageJ commented on plugins/out_stackdriver/stackdriver_conf.c:
```diff
@@ -286,6 +296,33 @@ struct flb_stackdriver *flb_stackdriver_conf_create(struct flb_output_instance *
+        }
+
+        if (!ctx->cluster_name || !ctx->cluster_location) {
+            flb_plg_error(ctx->ins, "Missing k8s_cluster_name "
```
lower_case? 'missing k8s_cluster_name'

JeffLuoo replied:
updated

edsiper commented:
LGTM.

cc: @courageJ @StevenYCChou

courageJ commented:
LGTM

StevenYCChou commented on plugins/out_stackdriver/stackdriver_conf.c:
```diff
@@ -286,6 +296,33 @@ struct flb_stackdriver *flb_stackdriver_conf_create(struct flb_output_instance *
         ctx->severity_key = flb_sds_create(tmp);
     }
 
+    if (flb_sds_cmp(ctx->resource, "k8s_container", 
```
[Optional] 

This means you can only use the output plugin for one type of resources. Is it your intention? If yes, I think I don't have concern.

I think this is fine for me as in some context you may only want to use fluent-bit to send specific resource type. I expect you can use single fluent-bit for multiple k8s resources though but it can be outside the range of this PR. in that case, you can  get the `cluster_location` and `cluster_name` from config without checking the resource and the configurations can be used for all k8s workloads.

JeffLuoo replied:
Hi Yen-Cheng I think the scope of this patch will support specific resource type first. This could be scaled up and support multiple resource types using one fluent-bit in the future. Thanks.

StevenYCChou replied:
SGTM.

JeffLuoo replied:
Good. Thanks

JeffLuoo commented:
Hi @edsiper Thanks for the review. And I am wondering that should I squash all of my 16 commits in to 1 commit before merging? 

Thanks.

edsiper commented:
ideally two commits:

- out_stackdriver: everything about what happened on plugins/out_stackdriver/*
- tests: runtime: out_stackdriver: any update on runtime tests for stackdriver
